### ISSUE TITLE
Build finished event now deprecated

### ISSUE THREAD

**Issue by baron1405** [gradle/gradle#20151]
Starting in Gradle 7.4, both the `Gradle.buildFinished` and `BuildListener.buildFinished` are deprecated. Apparently, it was not possible to get these methods to work with the configuration caching feature. The deprecation message does not list an alternative mechanism for being called when a build completes. In addition to this deprecation breaking one of my plugins, it seems like a serious regression to remove the ability to receive such a fundamental event as the completion of the build. Perhaps I am missing some other mechanism for receiving this event.

### Expected Behavior
Ability to register an event handler for build completion.

### Current Behavior
The buildFinished event has been deprecated with no apparent replacement.

### Context
Deprecation in Gradle 7.4.

**Comment by ljacomet** [comment]
Indeed, these are deprecated and planned to be removed.

Alternatives have been [documented](https://docs.gradle.org/current/userguide/upgrading_version_7.html#build_finished_events) but I can see that the Javadoc does not point to that information.

Is your build emitting a deprecation message? Or are you seeing the deprecation only in the context of the IDE when editing the script?

**Comment by ljacomet** [comment]
Gradle 7.5 will have more information on this deprecation and the alternatives that are configuration cache compatible.

**Comment by baron1405** [comment]
Thanks @ljacomet for the quick reply! I have my build treat deprecations as errors so that is why I see it. The IDE also shows the method as deprecated. Thanks for the pointer to the docs. I missed this while looking through the release notes. It will definitely help others to have a pointer to the alternatives in the Javadoc.

**Comment by thumannw** [comment]
@ljacomet The current docs (release notes) point to the Build Service feature with no further information. I checked the docs for 7.5 RC and nightly but did not find anything new regarding buildFinished. Will there be other alternatives in 7.5 as you mentioned above?

**Comment by bernhardkern** [comment]
@ljacomet: Still I do not get the documentation, even with the link you provided. Your link points to Build finished events -> configuration cache chapter ->  build services. 

Does it mean we have to implement a custom build service to solve the issue?

**Comment by aSemy** [comment]
> Indeed, these are deprecated and planned to be removed.
> 
> Alternatives have been [documented](https://docs.gradle.org/current/userguide/upgrading_version_7.html#build_finished_events) but I can see that the Javadoc does not point to that information.
> 
> Is your build emitting a deprecation message? Or are you seeing the deprecation only in the context of the IDE when editing the script?

Can you link directly to the documented alternatives? The link you provided isn't very clear. I eventually end up at the bottom of this page https://docs.gradle.org/current/userguide/build_services.html#operation_listener, but the description isn't clear.

> A build service can be used to receive events as tasks are executed. To do this, create and register a build service that implements [`OperationCompletionListener`](https://docs.gradle.org/current/javadoc/org/gradle/tooling/events/OperationCompletionListener.html). Then, you can use the methods on the [`BuildEventsListenerRegistry`](https://docs.gradle.org/current/javadoc/org/gradle/build/event/BuildEventsListenerRegistry.html) service to start receiving events.

What is that supposed to mean? What's a `BuildEventsListenerRegistry`?

**Comment by aSemy** [comment]
Okay I think I've figured it out. Here's a 'minimal' example you can copy into a `build.gradle.kts`. 

```kotlin
// build.gradle.kts

abstract class BuildListenerService :
    BuildService<BuildListenerService.Params>,
    org.gradle.tooling.events.OperationCompletionListener {

    interface Params : BuildServiceParameters

    override fun onFinish(event: org.gradle.tooling.events.FinishEvent) {
        println("BuildListenerService got event $event")
    }
}

val buildServiceListener = gradle.sharedServices.registerIfAbsent("buildServiceListener", BuildListenerService::class.java) { }

abstract class Services @Inject constructor(
    val buildEventsListenerRegistry: BuildEventsListenerRegistry
)

val services = objects.newInstance(Services::class)

services.buildEventsListenerRegistry.onTaskCompletion(buildServiceListener)
```

In the logs I can see the printed message

```none
> Task :service:vp:assemble UP-TO-DATE
> Task :assemble UP-TO-DATE
> Task :service:arc:processResources NO-SOURCE
> Task :service:ar:processResources UP-TO-DATE
> Task :service:ara:processResources UP-TO-DATE
BuildListenerService got event Task :service:vp:assemble UP-TO-DATE
BuildListenerService got event Task :assemble UP-TO-DATE
BuildListenerService got event Task :service:arc:processResources skipped
BuildListenerService got event Task :service:ar:processResources UP-TO-DATE
BuildListenerService got event Task :service:ara:processResources UP-TO-DATE
> Task :service:ti:kaptGenerateStubsKotlin UP-TO-DATE
BuildListenerService got event Task :service:ti:kaptGenerateStubsKotlin UP-TO-DATE
> Task :service:ac:kaptGenerateStubsKotlin UP-TO-DATE
BuildListenerService got event Task :service:ac:kaptGenerateStubsKotlin UP-TO-DATE
> Task :service:ti:kaptKotlin UP-TO-DATE
BuildListenerService got event Task :service:ti:kaptKotlin UP-TO-DATE
> Task :service:ti:compileKotlin NO-SOURCE
BuildListenerService got event Task :service:ti:compileKotlin skipped
> Task :service:ti:compileJava NO-SOURCE
BuildListenerService got event Task :service:ti:compileJava skipped
> Task :service:ti:processResources NO-SOURCE
```

**Comment by ingokegel** [comment]
@aSemy Thank you!

I have tried this too, but `BuildEventsListenerRegistry.onTaskCompletion` is only for tasks, not for the entire build. There would have to be a method ` BuildEventsListenerRegistry.onProjectCompletion`, then it would work.

@eskatos Can you add something like the above for Gradle 7.6 or explain if there is a different way to notify the build service that the entire build has completed?

**Comment by SimonCJacobs** [comment]
I just added this to my Kotlin script and it seems to be working ok. Pleased to hear of any improvements.
```
val buildEvents: BuildEventsListenerRegistry // obtain by injection into appropriate object, eg per @aSemy

project.gradle.taskGraph.whenReady {
   project.gradle.sharedServices.registerIfAbsent("buildFinished", BuildFinishedService::class.java) {
      parameters.lastTaskPath = project.gradle.taskGraph.allTasks.last().path
   }
      .let { buildEvents.onTaskCompletion(it) }
}

abstract class BuildFinishedService : BuildService<BuildFinishedService.Parameters>, OperationCompletionListener {
   interface Parameters : BuildServiceParameters {
      var lastTaskPath: String
   }

   override fun onFinish(event: FinishEvent?) {
      (event as TaskFinishEvent).let {
         if (event.descriptor.taskPath == parameters.lastTaskPath)
            buildFinishedAction()
      }
   }

   private fun buildFinishedAction() {
      // Do something exciting
   }
}
```

**Comment by Crazyjavahacking** [comment]
The previously mentioned code snippet does not cover all use cases.

E.g. we have a code that gets the access to build failure's exception stack trace. With build service, it's not possible.

Also if a code in project.getGradle().projectsEvaluated() throws exception, it will not be catched by the OperationCompletionListener (tested with Gradle 7.5).

**Comment by adrianvintu** [comment]
My `gradle.buildFinished` is a single line of code playing an mp3 when the build is done.

In the new version, will I need to write 20 lines (of code I don't understand) in order to achieve the same thing?

**Comment by ThadHouse** [comment]
> In the context of using the configuration cache you can use an `AutoCloseable` build service as an approximation of the end of the build. 
> 

Is this behavior going to be guaranteed in the future? Where an AutoCloseable service isn't closed until the end of the build? I could see Gradle adding an optimization where it will close the service after all tasks that declare service usage finish. Which would break this assumption.

**Comment by ThadHouse** [comment]
Going through this again, using the BuildEventsListenerRegistry is not enough for my use case. If an exception caused the build failure, I need the actual exception that caused the failure. And the registry does not provide that. And this was using fully public types to do this in the past, as BuildResult would provide a Throwable, so that is a full regression.

**Comment by eskatos** [comment]
@ThadHouse, see https://github.com/gradle/gradle/issues/17659

**Comment by maguro** [comment]
Respectfully, what the heck?  Speaking to those who thought that deprecating `buildFinished()` was a good idea, possibly, in favor of a `BuildEventsListenerRegistry` implementation: can anyone think

**Comment by bamboo** [comment]
Hi everyone, [this public specification](https://docs.google.com/document/d/1nS1JTwdCMQN_0WaaFrSUxoQOj3zOqmKmXxC6k0OXUR0) describes how we're planning to improve the situation in 8.1.
Let us know what you think.

**Comment by bamboo** [comment]
Implemented in #17659

**Comment by asarkar** [comment]
@bamboo Is there any documentation with examples showing how to use the new APIs to solve the problem this ticket was for?

**Comment by ingokegel** [comment]
@asarkar 

https://docs.gradle.org/current/userguide/dataflow_actions.html

It works, but the concepts you have to understand are quite a handful for such a simple task. I wonder if one could develop some sort of convenience method that is close to the original buildFinished listener.

**Comment by asarkar** [comment]
@ingokegel I've read that already, but it's so severely constricted it seems to be written for Twitter in mind. I know no more after reading it than I knew before.
[severely constricted](https://docs.gradle.org/current/userguide/dataflow_actions.html#lifecycle_event_providers)

**Comment by ingokegel** [comment]
@asarkar It is indeed very complicated. On the `FlowScope` that you get from the injected `getFlowScope()` method, you call `always()` (so it is "always" part of the task graph which is currently the only option) with a lambda that calls `getFlowProviders().buildWorkResult.map[ /* configure FlowActionSpec */ }` on the injected `FlowProvider` which runs your code in the `FlowActionSpec` lambda. There you wire a Gradle Properties field in your `FlowParameters` object with a provider that uses the action build result, possibly just a boolean.

`buildWorkResult` (currently the only available flow provider) becomes available which is when the build is finished.  When all properties of the `FlowAction` are available it is then executed automatically (this is somewhat strange and not explained in the documentation). It is a lot harder to access any other state and you have to do it with the `FlowParameters` of the `FlowAction`, using Gradle `Property` fields in the parameter object but remember that the flow action will only be triggered once all of the properties on the `FlowAction` become available. 

Looking at the code it is very difficult to understand what is happening and I really wonder if all of this is necessary just to work with the configuration cache.

**Comment by alexeiartsimovich** [comment]
Hello!
I can track the end of builds using `getFlowProviders().buildWorkResult.map{ /* configure FlowActionSpec */ }` but is there any way to know when the build started? I want to know the entire build duration.

**Comment by JojOatXGME** [comment]
Hi, I was trying to replace [the usage of `buildFinished` in the *nix-idea* plugin](https://github.com/NixOS/nix-idea/blob/57132bec7080cdb554f8278983967dc3013bc319/build.gradle.kts#L80-L102). The code checks if the build failed for a particular reason, and if so, provides some guidance for the user. Here is what I came up with for now. I have to say that I don't like that the parameters implicitly define when the action is executed. Having real lifecycle hooks would seem more intuitive and less error-prone to me. Is there a reason for modeling a new dynamic execution graph next to tasks, instead of simply exposing a method which allows adding `BUILD_PHASE` listeners next to `BuildEventsListenerRegistry.onTaskCompletion`?

<details>
<summary>Show potential replacement for code in nix-idea ...</summary>

```kotlin
import org.jetbrains.intellij.tasks.RunIdeBase

// I haven't found a way to access services (like FlowScope) from script plugins,
// so I created this nested plugin class.
apply<DelegatePlugin>()
abstract class DelegatePlugin : Plugin<Project> {
    @get:Inject
    protected abstract val flowScope: FlowScope

    @get:Inject
    protected abstract val flowProviders: FlowProviders

    override fun apply(target: Project) {
        val jbr = target.providers.of(JbrSource::class.java) {
            parameters.path.set(target.file("jbr/bin/java"))
        }.orNull

        if (jbr == null) {
            flowScope.always(JbrGuidance::class) {
                parameters.project.set(target)
                parameters.buildResult.set(flowProviders.buildWorkResult)
            }
        } else {
            target.tasks.withType<RunIdeBase> {
                projectExecutable.set(jbr)
            }
        }
    }
}

abstract class JbrSource : ValueSource<String?, JbrSource.Parameters> {
    interface Parameters : ValueSourceParameters {
        @get:Input
        val path: Property<File>
    }

    override fun obtain(): String? {
        return parameters.path.get().takeIf { it.exists() }?.toString()
    }
}

abstract class JbrGuidance : FlowAction<JbrGuidance.Parameters> {
    private val regex = Regex("""\.gradle/.*/jbr/.*/java\b""")

    interface Parameters : FlowParameters {
        @get:Input
        val project: Property<Project>

        @get:Input // Unused, but tells Gradle that this action shall be executed at the end of the build
        val buildResult: Property<BuildWorkResult>
    }

    override fun execute(parameters: Parameters) {
        //val logger = org.slf4j.LoggerFactory.getLogger("jbr-guidance")
        val project = parameters.project.get()
        val logger = project.logger
        val taskGraph = project.gradle.taskGraph
        for (task in taskGraph.allTasks) {
            if (task.project == project && task is RunIdeBase &&
                task.state.failure?.cause?.message?.contains(regex) == true
            ) {
                logger.error("""
                    |
                    |! Info for users on NixOS:
                    |!
                    |! The JetBrains Runtime (JBR) downloaded by Gradle is not compatible with NixOS.
                    |! You may run the ‘:jbr’ task to configure the runtime of <nixpkgs> instead.
                    |! Alternatively, you may run the following command within the project directory.
                    |!
                    |!   nix-build '<nixpkgs>' -A jetbrains.jdk -o jbr
                    |!
                    |! This will create a symlink to the package jetbrains.jdk of nixpkgs at
                    |! ${'$'}projectDir/jbr, which is automatically detected by future builds.
                    """.trimMargin())
                break
            }
        }
    }
}

```

</details>

**Comment by samoylenkodmitry** [comment]
Is there any example project showing how to replace 
```
	gradle.buildFinished {
...
	}
```
to the working solution with all project dirs/paths that are needed provided? Quick stuck with broken build.

**Comment by TWiStErRob** [comment]
@samoylenkodmitry there's a working example in [the Gradle repo](https://github.com/gradle/gradle/tree/master/subprojects/docs/src/snippets/dataflowActions/playSound/groovy) which is used to generate [the docs](https://docs.gradle.org/current/userguide/dataflow_actions.html).

(Untested suggestion: You don't necessarily need the full ceremony of a Settings plugin, you can copy the body of `apply` (along with the two other classes into `settings.gradle.kts`, and get the two injected fields imperatively via `serviceOf<T>()` in Kotlin DSL.)

**Comment by Fermiz** [comment]
Thanks to @SimonCJacobs's demo and serviceOf<T>() in Kotlin DSL mentioned by @TWiStErRob, I made it after half a day's spike. To save time, here is my full code snippets to print the time-consuming statistics of each task: 

```
import org.gradle.tooling.events.OperationCompletionListener
import org.gradle.tooling.events.FinishEvent
import org.gradle.kotlin.dsl.support.serviceOf
import org.gradle.tooling.events.OperationResult
import org.gradle.tooling.events.FailureResult
import org.gradle.tooling.events.SkippedResult

abstract class StatisticsService : BuildService<StatisticsService.Parameters>,
    OperationCompletionListener {

    interface Parameters : BuildServiceParameters {
        var lastTask: String
        var timeStatistics: MutableMap<String, Long>
    }

    override fun onFinish(event: FinishEvent) {
        val taskName = event.descriptor.name
        val operationResult = event.result
        val durationSeconds = TimeUnit.MILLISECONDS.toSeconds(operationResult.endTime - operationResult.startTime)

        parameters.timeStatistics.putIfAbsent(taskName, durationSeconds)
        println("[$taskName] time elapsed: ${durationSeconds}s")

        if (operationResult is FailureResult || taskName == parameters.lastTask) {
            println("//// Task Time Statistics ////")
            parameters.timeStatistics.toList().sortedByDescending { (_, duration) -> duration }.toMap()
               .forEach { (task, duration) -> println("[${task.padEnd(32)}] : ${duration}s") }
        }
    }
}

gradle.taskGraph.whenReady {
    val lastTask = gradle.taskGraph.allTasks.last().path

    val timeStatistics: MutableMap<String, Long> = mutableMapOf()

    val demoListener = gradle.sharedServices.registerIfAbsent(
        "demoListener ", StatisticsService::class.java
    ) {
        parameters.lastTask = lastTask
        parameters.timeStatistics = timeStatistics
    }

    gradle.serviceOf<BuildEventsListenerRegistry>().onTaskCompletion(demoListener)
}
```
Hope it would be useful for the new beginers who is looking for the gradle hooks alternative on the latest version.

**Comment by TWiStErRob** [comment]
_Sorry for the off-topic, but would like to point out some potential pitfalls related to @Fermiz's code. I recommend continuing discussion in [Gradle Community Slack](https://gradle.com/slack-invite)._

For you "build-end" detection it's probably better to use the `FlowScope.always` event, even if it's a bit more code, because the way the "last task" is used above may not always be correct.

<details><summary>For example in AGP</summary>

We have `assembleDebug` and `assembleRelease`, both building the same-ish thing in `--parallel`, but `assembleRelease` contains a minification step, which makes it overall longer. If the command is `gradlew assembleRelease assembleDebug`, the last command is likely `assembleDebug`, but it's very likely `minifyReleaseWithR8` (very long task) and therefore `assembleRelease` (which depends on R8) will complete later than than the "last task" picked from the graph.

</details>

---

Also with `--continue` one can have multiple failing tasks in the same build.

<details><summary>For example</summary>

In a multi-module project, `detekt` (or similar static analysis tool) can run and fail on each module resulting in summary statistics printed many times.

</details>

---

Both of these could result in mis-information from the stats, missing long tasks from the table; or if the code is not printing, duplicate processing of potentially wrong data.

**Comment by JojOatXGME** [comment]
@Fermiz, a `BuildService` can just implement `AutoCloseable` to detect the end of the build. The only limitation compared to the `FlowAction` is that you will not get the overall build result. However, you can get the result of each individual task, as you are already doing in your `StatisticsService`. Here is an example where I used `AutoCloseable` in a build service to replace `gradle.buildFinished`.

https://github.com/NixOS/nix-idea/blob/f25cb9b9714bfd05fc71c8490ea4f4f84ee72e79/gradle/plugins/src/main/kotlin/local.jbr-guidance.gradle.kts#L70-L88

Note that a build may still fail, even if there is not a single failing task. For example, the serialization of task inputs may fail when the configuration cache is enabled. I don't know why Gradle doesn't just provide a listener interface for build services to be called on build completion, similar to `BuildEventsListenerRegistry.onTaskCompletion`.

**Comment by LinqLover** [comment]
`gradle.serviceOf()` is internal not part of the official API, so the solution by @Fermiz seems at least not completely robust. 😕

**Comment by TWiStErRob** [comment]
@LinqLover while that is not part of the official API it's quite trivial to achieve its behaviour, see for a similar example see https://docs.gradle.org/current/userguide/service_injection.html#filesystemoperations

**Comment by matejonnet** [comment]
Using `finalizedBy` to make sure the task is always in the execution graph and `mustRunAfter` to keep the order, seems to work.

The reportCompletion task (always run the last) is defined to finalize the taskA (first task) and at the same time it is defined that it must run after all tasks.

Test example:
```
// Test file to verify task dependency behavior
// Run with: gradle testScenario1
// Or: gradle testScenario2

task taskA {
    doLast {
        println "✓ TaskA executed"
    }
}

task taskB(dependsOn: taskA) {
    doLast {
        println "✓ TaskB executed"
        throw new RuntimeException("TaskB intentionally fails!")
    }
}

task taskC(dependsOn: taskB) {
    doLast {
        println "✓ TaskC executed (should NOT run if taskB fails)"
    }
}

task taskD {
    doLast {
        println "✓ TaskD executed (runs via finalizedBy)"
    }
}

// TaskD always runs via finalizedBy on taskB
taskB.finalizedBy taskD

// Test scenario 1: TaskB fails, TaskC is skipped, but reportCompletion should still run
task testScenario1(dependsOn: taskC) {
    group = 'test'
    description = 'Test scenario where taskB fails and taskC is skipped'
}

// Test scenario 2: All tasks succeed
task taskE(dependsOn: taskA) {
    doLast {
        println "✓ TaskE executed (no failure)"
    }
}

task testScenario2(dependsOn: taskE) {
    group = 'test'
    description = 'Test scenario where all tasks succeed'
}

// Define reportCompletion AFTER all other tasks
task reportCompletion {
    doLast {
        println "✓ REPORT COMPLETION EXECUTED"
        
        def failedTasks = gradle.taskGraph.allTasks.findAll { task ->
            task.state.failure != null
        }
        
        if (!failedTasks.isEmpty()) {
            println "Failed tasks: ${failedTasks.collect { it.name }.join(', ')}"
        } else {
            println "All tasks succeeded"
        }
    }
}

// Programmatically set mustRunAfter for all other tasks after all tasks are defined
afterEvaluate {
    def allOtherTasks = tasks.findAll { it.name != 'reportCompletion' }
    reportCompletion.mustRunAfter allOtherTasks
    println "reportCompletion.mustRunAfter: ${allOtherTasks.collect { it.name }.join(', ')}"
}

// Attach reportCompletion to taskA (which always runs)
taskA.finalizedBy reportCompletion

defaultTasks 'testScenario1'
```
